syncthing-bep/src/main/kotlin/net/syncthing/java/bep/BlockPuller.kt

```diff
@@ -14,145 +15,175 @@
+    ): InputStream {
+        return runBlocking {
+            pullFileCoroutine(fileInfo, progressListener)
+        }
+    }
+
+    suspend fun pullFileCoroutine(
+            fileInfo: FileInfo,
+            progressListener: (progress: Double, progressMessage: String) -> Unit = { _, _ -> }
+    ): InputStream {
         val fileBlocks = indexHandler.waitForRemoteIndexAcquired(connectionHandler)
                 .getFileInfoAndBlocksByPath(fileInfo.folder, fileInfo.path)
                 ?.value
                 ?: throw IOException("file not found in local index for folder = ${fileInfo.folder} path = ${fileInfo.path}")
         logger.info("pulling file = {}", fileBlocks)
-        NetworkUtils.assertProtocol(connectionHandler.hasFolder(fileBlocks.folder), {"supplied connection handler $connectionHandler will not share folder ${fileBlocks.folder}"})
-        val error = AtomicReference<Exception>()
-        val fileDownloadObserver = object : FileDownloadObserver() {
-
-            private fun receivedData() = (blocksByHash.size * BlockPusher.BLOCK_SIZE).toLong()
+        NetworkUtils.assertProtocol(connectionHandler.hasFolder(fileBlocks.folder), { "supplied connection handler $connectionHandler will not share folder ${fileBlocks.folder}" })
 
-            private fun totalData() = ((blocksByHash.size + missingHashes.size) * BlockPusher.BLOCK_SIZE).toLong()
+        val totalTransferSize = fileBlocks.blocks.distinctBy { it.hash }.longSumBy { it.size.toLong() }
 
-            override fun progress() = if (isCompleted()) 1.0 else receivedData() / totalData().toDouble()
+        val blockTempIdByHash = Collections.synchronizedMap(HashMap<String, String>())
 
-            override fun progressMessage() = (Math.round(progress() * 1000.0) / 10.0).toString() + "% " +
-                    FileUtils.byteCountToDisplaySize(receivedData()) + " / " + FileUtils.byteCountToDisplaySize(totalData())
+        try {
+            var receivedData = 0L
+            val reportProgressLock = Object()
 
-            override fun isCompleted() = missingHashes.isEmpty()
+            fun updateProgress(newReceivedDataSize: Long) {
+                synchronized(reportProgressLock) {
+                    receivedData += newReceivedDataSize
 
-            override fun inputStream(): InputStream {
-                    NetworkUtils.assertProtocol(missingHashes.isEmpty(), {"pull failed, some blocks are still missing"})
-                    val blockList = hashList.map { blocksByHash[it] }.toList()
-                    return SequenceInputStream(Collections.enumeration(blockList.map { ByteArrayInputStream(it) }))
-                }
+                    val progress = receivedData / totalTransferSize.toDouble()
+                    val progressMessage = (Math.round(progress * 1000.0) / 10.0).toString() + "% " +
+                            FileUtils.byteCountToDisplaySize(receivedData) + " / " + FileUtils.byteCountToDisplaySize(totalTransferSize)
```

Comment:
It's strange that display code (how the string is formatted) lives in the download loop. Can't you just pass the bytes to the progress listener, and do string manipulation in the UI code?

Reply:
This string is never used in the UI. However, I replaced this with an status object which is passed to the UI (which makes it easier to add new fields).